ccvm/api.go

```diff
@@ -0,0 +1,314 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"os"
+
+	"github.com/intel/ccloudvm/types"
+)
+
+// ServerAPI exposes an RPC based API to the cccloudvm client.  This API can
+// be used to perform almost all ccloudvm actions.  The API has the following
+// structure.  All APIs calls are asynchronous.  The client initiates a request
+// and receives a transaction id.  At this stage the command is not processed.
+// It's just scheduled.  The client can cancel the command by passing the transaction
+// ID to the Cancel function or by calling the corresponding Result function
+// to which he must also pass the transaction ID.  Some Result functions, e.g.,
+// CreateResult need to be called multiple times.  The values returned from
+// such functions will indicate whether they need to be called again or not.
+// And example call sequence would be.
+//
+// Start -> id = 2
+// StartResult(2)
+//
+type ServerAPI struct {
+	signalCh chan os.Signal
+	actionCh chan interface{}
+}
+
+func (s *ServerAPI) sendStartAction(fn func(context.Context, *service, chan interface{}), id *int) {
+	action := startAction{
+		action:  fn,
+		transCh: make(chan int),
+	}
+	s.actionCh <- action
+	*id = <-action.transCh
+}
+
+func (s *ServerAPI) voidResult(id int, reply *struct{}) error {
+	result := getResult{
+		ID:  id,
+		res: make(chan interface{}),
+	}
+
+	s.actionCh <- result
+	r := <-result.res
+	if v, ok := r.(error); ok {
+		return v
+	}
+
+	resultCh := r.(chan interface{})
+	err, _ := (<-resultCh).(error)
+	*reply = struct{}{}
+
+	s.actionCh <- completeAction(id)
+
+	return err
+}
+
+// Cancel can be used to cancel any command that has been issued but not
+// yet completed.
+func (s *ServerAPI) Cancel(arg int, reply *struct{}) error {
+	fmt.Printf("Cancel(%d) called\n", arg)
+	s.actionCh <- cancelAction(arg)
+	*reply = struct{}{}
+	return nil
+}
+
+// Create initiates a new instance creation request using the arguments provided by the
+// args parameter. The value pointed to by id is set to the transaction ID of the request
+// if no error occurs.
+func (s *ServerAPI) Create(args *types.CreateArgs, id *int) error {
+	fmt.Printf("Create %+v called\n", *args)
+
+	s.sendStartAction(func(ctx context.Context, svc *service, resultCh chan interface{}) {
+		svc.create(ctx, resultCh, args)
+	}, id)
+
+	fmt.Printf("Transaction ID %d\n", *id)
+	return nil
+}
+
+// CreateResult blocks until information about the instance creation request has
+// been received.  This information could be an error, signalling that the
+// request has failed or a types.CreateResult.  CreateResult should be called continually
+// until res.Finished == true.  If successful, the final types.CreateResult returned will
+// have the its Finished field set to true and its Name field set to the name of the instance.
```

Comment:
s/have the its Finished/have its Finished/

Reply:
Thanks.  Will fix.